### DISCUSSION TITLE
Unwanted blank namespace appearing in list child element tag

### DISCUSSION THREAD

**Discussion by alucasa** [pdvrieze/xmlutil#252] [Q&A] [answered]
I am using `core-jvm` and `serialization-jvm` versions `0.86.3` to try and remove manual XML serialization and deserialization from a project. I am running into an issue where a child element nested several layers below the root element is serializing with an unwanted blank namespace `xmlns=""`. Otherwise, everything is serializing as expected.

I am trying to serialize a `CreateHouseRequest` as the root element where the namespace should be specified. Specifically, the unwanted blank namespace is appearing on the child element `FloorDetail` in the list `floorDetails` (member of `HouseModel`). However, if I explicitly specify the the namespace in the `@XmlSerialName` annotation for `fancyHouse` in `CreateHouseRequest` then the blank namespace appears there on the `House` element, instead of in the children of `floorDetails`. 

I have tried using `@XmlNamespaceDecl` on the root element (removing namespace from the @XmlSerialName annotation), but this generates a namespace with the unwanted prefix `n1` even though I am using a single string value (no `;` separations or prefixes specified with `=`).

I'm unable to identify where I'm going wrong, and am looking for some help to point me in the right direction. I'd like it so that only the root element `CreateHouseRequest` has the namespace specified to MyNamespaceUris.HouseURI, and no other elements inherit or serialize with a namespace. 

XML configuration:
```
XML(SerializerModules.houseRequestModule) {
    recommended_0_86_3 {
        indent = 2
        autoPolymorphic = true
    }
}
```

SerializersModule:

```
houseRequestModule  = SerializersModule {
  polymorphic(HouseRequest::class) {
    subclass(CreateHouseRequest::class, CreateHouseRequestRequest.serializer())
    subclass(UpdateHouseRequest::class, UpdateHouseRequest.serializer())
  }
```

Relevant Classes: 

```
@Serializable
@XmlSerialName("HouseCreationRequest", MyNamespaceUris.HouseURI)
data class CreateHouseRequest(
    @XmlSerialName(House.TAG_NAME)
    private val fancyHouse: FancyHouse,
) : HouseRequest

interface House {
  companion object {
    const val TAG_NAME = "House"
  }
}

@Serializable(with = FancyHouseSerializer::class)
data class FancyHouse(
    @XmlSerialName(House.TAG_NAME)
    val houseModel: HouseModel,
) : House


@Serializable
data class SimpleHouse(
    @Serializable(with = HouseIdSerializer::class)
    @XmlSerialName("HouseId")
    private val houseId: HouseId,
) : House


@Serializable
data class HouseModel(
    @XmlSerialName("Name") val name: String,
    @XmlSerialName("Description")
    private val description: String,
    @Serializable(with = LatLongSerializer::class)
    @XmlSerialName("Location")
    private val location: LatLong,
    @XmlSerialName("FloorDetails")
    @XmlChildrenName("FloorDetail")
    val bedrooms: List<FloorDetail>,
)


@Serializable
data class FloorDetail(
    @Serializable(with = PercentSerializer::class)
    @XmlSerialName("First_Percent")
    private val firstPercent: Percent,
    @Serializable(with = PercentSerializer::class)
    @XmlSerialName("Second_percent")
    private val secondPercent: Percent,
    @XmlSerialName("Bathrooms")
    @XmlChildrenName("Bathroom")
    private val inverters: List<Bathroom>,
    @XmlSerialName("Bedrooms")
    @XmlChildrenName("Bedroom")
    private val pvArrays: List<Bedroom>,
)
``` 

FancyHouseSerializer:

```
class FancyHouseSerializer : KSerializer<FancyHouse> {
    @OptIn(ExperimentalSerializationApi::class)
    override val descriptor: SerialDescriptor =
        SerialDescriptor(
            serialName = House.TAG_NAME,
            original = HouseModel.serializer().descriptor,
        )

    override fun serialize(
        encoder: Encoder,
        value: FancyHouse,
    ) = encoder.encodeSerializableValue(HouseModel.serializer(), value.houseModel)

    override fun deserialize(decoder: Decoder): FancyHouse {
        val hardwareModel = decoder.decodeSerializableValue(HouseModel.serializer())
        return FancyHouse(hardwareModel)
    }
}

```

Example of what I'm seeing where the unwanted blank xmlns attribute is serialized on `FloorDetail`:

```
<HouseCreationRequest xmlns="http://mynamespace.com">
  <House Name="Test site" Description="Test site description">
    <Location Latitude="34.65" Longitude="-119.1"/>
    <FloorDetails>
      <FloorDetail  xmlns="" First_Percent="17.0" SecondPercent="85.0">
        <Bathrooms>
          <Bathroom .... />
        </Bathrooms>
        <Bedrooms>
          <Bedroom... />>
            ...
```

**Comment by alucasa** [comment]
I was able to get around this issue (no longer seeing blank namespace in `FloorDetail` child) by removing use of the `@XmlChildrenName` annotation and instead use a custom serializer for lists of non-primitive types (below). However, if this can be achieved without the hand rolled list serializer that would be preferred. 

```
open class ListSerializer<T>(
    private val elementName: String,
    private val childSerializer: KSerializer<T>,
) : KSerializer<List<T>> {
    override val descriptor: SerialDescriptor =
        buildClassSerialDescriptor(elementName) {
            element(elementName, childSerializer.descriptor)
        }

    override fun serialize(
        encoder: Encoder,
        value: List<T>,
    ) {
        encoder.encodeStructure(descriptor) {
            for (item in value) {
                encodeSerializableElement(descriptor, 0, childSerializer, item)
            }
        }
    }

    override fun deserialize(decoder: Decoder): List<T> {
        val items = mutableListOf<T>()
        decoder.decodeStructure(descriptor) {
            while (true) {
                val index = decodeElementIndex(descriptor)
                if (index == CompositeDecoder.DECODE_DONE) break
                items.add(decodeSerializableElement(descriptor, index, childSerializer))
            }
        }
        return items
    }
}
```

**Comment by pdvrieze** [comment]
I'm assuming that you want `FancyHouse` to be actually in the `"http://mynamespace.com"` namespace, not in the default/null namespace. If `FancyHouse` is in the default namespace then the namespace attribute is correct.
For `XmlChildrenName` you can specify the namespace attribute and in that case it should be picked up. However, the defaulting mechanism is broken (probably because default annotation arguments didn't work reliably until K2 so it is relatively untested).

If you give the namespace explicitly it works (in my test case), the problem is that it defaults to the empty namespace (not the one of the parent tag as is the behaviour of `XmlSerialName).

**Reply by alucasa** [reply]
I'd like it so that only the root element `HouseCreationRequest` has a namespace attribute and that no other element has one. From your comment it doesn't sound like this is possible using `@XmlChildrenName` to handle lists, and that I'll have to stick with my custom serializer.

What I'd like to see (`FloorDetail` no longer has a blank namespace or any namespace attribute)

```
<HouseCreationRequest xmlns="http://mynamespace.com">
  <House Name="Test site" Description="Test site description">
    <Location Latitude="34.65" Longitude="-119.1"/>
    <FloorDetails>
      <FloorDetail  First_Percent="17.0" SecondPercent="85.0">
        <Bathrooms>
          <Bathroom .... />
        </Bathrooms>
        <Bedrooms>
          <Bedroom... />>
            ...
```

**Reply by pdvrieze** [reply, marked as answer]
I think you are confused as to how it works. In your example you declare that HouseCreationRequest lives in the "http://mynamespace.com" namespace with the default prefix. The `xmlns` namespace declaration just specifies the prefix/namespace mapping. All tags have a namespace, based on the prefix of the element (also if empty). So `House`, `Location`, `FloorDetails`, `FloorDetail`, `Bathrooms` and `Bathroom` are all in that namespace, but because you use the default namespace this is not possible. The bug with `@XmlChildrenName` is that instead of inheriting the namespace it uses the empty namespace. That is the reason you see the `xmlns=""` attribute on `FloorDetail` because it is now in the default/empty namespace. This namespace must use the empty prefix (per the XML standard), and that prefix needs to be overridden for that element (it's children will now inherit the default namespace).

As a sidenote, attributes without prefix are always in the null/default namespace, and practice is to not use namespaced attributes for internal (not imported) attributes.

So, try it out, specify namespace explicitly in `@XmlChildrenName` and you will not see an `xmlns` attribute because the writer will reuse the existing namespace/prefix mapping on the parent (root) element.

What you want is equivalent to:
```xml
<ns:HouseCreationRequest xmlns:ns="http://mynamespace.com">
  <ns:House Name="Test site" Description="Test site description">
    <ns:nsLocation Latitude="34.65" Longitude="-119.1"/>
    <ns:FloorDetails>
      <ns:FloorDetail  First_Percent="17.0" SecondPercent="85.0">
        <ns:Bathrooms>
          <ns:Bathroom .... />
        </ns:Bathrooms>
        <ns:Bedrooms>
          <ns:Bedroom... />
```

If you don't specify the namespace attribute for the `FloorDetail` child/element (or set it to the empty namespace) you get:
```xml
<ns:HouseCreationRequest xmlns:ns="http://mynamespace.com">
  <ns:House Name="Test site" Description="Test site description">
    <ns:nsLocation Latitude="34.65" Longitude="-119.1"/>
    <ns:FloorDetails>
      <FloorDetail  First_Percent="17.0" SecondPercent="85.0">
        <Bathrooms>
          <Bathroom .... />
        </Bathrooms>
        <Bedrooms>
          <Bedroom... />
```

This should make it a bit clearer what is happening (but you want to use the empty prefix which reduces the clarity a bit but is just as valid), and any XML consumer should treat both the same (prefixes are arbitrary, hence automatic generation of them being valid).

**Reply by alucasa** [reply]
Thank you for the detailed explanation, it did indeed clarify what is happening. I was able to confirm serialization works as expected by deleting the custom list serializers and specifying the namespace in `@XmlChildrenName`. Much appreciated!
